astropy/io/ascii/cds.py

```diff
@@ -164,14 +208,302 @@ def get_cols(self, lines):
                     raise ValueError(f'Line "{line}" not parsable as CDS header')
 
         self.names = [x.name for x in cols]
-
         self.cols = cols
 
+    def __strFmt(self, string):
+        """Return argument formatted as string."""
+        if string is None:
+            return ""
+        else:
+            return string
+
+    def __get_type(self, col):
+        """Return column type."""
+        if col.dtype.name.startswith("i"):
+            return int
+        elif col.dtype.name.startswith("f"):
+            return float
+        elif col.dtype.name.startswith("u"):
+            return int
+        elif col.dtype.name.startswith("s"):
+            return str
+
+    def __splitFloatFormat(self, value, fmt):
+        """ get float format
+        value (IN) the float value
+        fmt (out) [size: length of string containing the float value,
+                   prec: precision of the column values, sum of dec and ent,
+                   dec: number of digits places after decimal point,
+                   ent: number of digits places before decimal point,
+                   sign: bool, is float value signed]
+        return True has scientific notation
+        """
+        regfloat = re.compile(r"""(?P<sign> [+-]*)
+                                  (?P<int> [^eE.]+)
+                                  (?P<deciPt> [.]*)
+                                  (?P<decimals> [0-9]*)
+                                  (?P<exp> [eE]*-*)[0-9]*""",
+                             re.VERBOSE)
+        mo = regfloat.match(value)
+
+        if mo is None:
+            raise Exception(value + " is not a float number")
+
+        fmt[0] = len(value)
+        if mo.group('sign') != "":
+            fmt[4] = True
+        else:
+            fmt[4] = False
+        fmt[2] = len(mo.group('int'))
+        fmt[3] = len(mo.group('decimals'))
+        fmt[1] = fmt[2] + fmt[3]
+
+        if mo.group('exp') != "":
+            # scientific notation
+            return True
+        return False
+
+    def columnFloatFormatter(self, col):
+        """
+        String formatter for a column containing Float values.
+        """
+        if col.hasNull: # optimized (2x more speed)
```

Comment:
This is premature optimization. Columns are numpy arrays and `col.max()` will do. To account for columns that are fully masked you can do something like:
```
col.max = col.max()
if np.is_masked(col.max):
    col.max = None
```
[At least I think I understand the logic of this bit of code. I'm not sure. But that fact alone shows that it might be too complex.]

Reply:
I think the way this part of the code works is something as follows.

- The max/min values have to be found in column containing Null values. 
- If the min/max values themselves are masked, then we want to assign `None` to them instead of the random masked value (or NaN).
- So, the null values in the column are filled with a large negative number `-999999` (a large positive number `999999`). If all the values in the column are Null values, then there won't be any number greater (smaller) than this filled value, and thus, the `col.max` (`col.min`) value will be set as `None`.

I looked up this page for more info: https://docs.astropy.org/en/stable/table/masking.html

I am not sure of the way in which these steps above optimize the process, since checking Min/Max will anyway depend on how well numpy functions are optimized. Maybe the default Min/Max checking is time consuming for columns with a very large number of values, in the worst case scenario and the above steps ensure lesser time?

Reply:
The numpy code itself is probably written in C (much on numpy internal are), so I'm very surprised if this Python is much faster. Maybe it is. The point is that we first want to get a code that working and that we understand. "I think this works something as follows" is not good enough :-)  So, I suggest to use a very simple solutions (like my three lines instead of > dozen lines with two different code path) at this point. If there is an if..else we need to write tests for both cases etc. It's just not worth it unless we **know** that this part of the code is a bottle neck for speed.

Reply:
Okay. I have already implemented that in [__set_column_val_limits()](https://github.com/astropy/astropy/pull/11897/files#diff-f1aebe08fb4f3dd370b7dec776d172228663c8a0ae39b8115e9f678b8199e42bR249-R259) function.
The optimization is probably good only for extremely large tables. The current implementation poses no problems for now.